### ISSUE TITLE
Provide contribution development installation guidelines

### ISSUE THREAD

**Issue by humanfactors** [easystats/easystats#46]
This is sort of a question, but I do think any answers would be well placed in a wiki or some form of contribution guidelines documentation. In short, I'm actually not sure the best way to set setup `easystats` in a development environment. There's really two parts to this:

[✔ - **Not issue for me**] Firstly, should a new user want to help contribute, my understanding is that they would need to first install all easystats packages on the dev branch

```
install_easystats_dev <- function(){
  devtools::install_github(c("easystats/insight",
                             "easystats/bayestestR",
                             "easystats/performance",
                             "easystats/parameters",
                             "easystats/effectsize",
                             "easystats/correlation",
                             "easystats/estimate",
                             "easystats/see",
                             "easystats/report"),
                           ref = "dev",
                           upgrade = "never")
}
```
This isn't documented on the main page, and while it's certainly not a big issue, the instructions in the `README.md` imply to me that you'd first ` devtools::install_github("easystats/easystats")`  and then call `install_easystats_dev`, but this seems excessive of course. I actually should note that I have had a few troubles installing with `install_easystats_dev` when I've already had dev packages installed, but I haven't really had time to debug

[❓ **My confusion**] For packages that I contribute too, or modify, I obviously have my own fork of the package. I then develop a branch on this, and make sure I am regularly updating the other packages to the dev branch. To be honest, this actually works fine, but it does make it hard to scale up to keeping track of developing multiple packages in this expanding and quite large ecosystem.

I'm guessing there's an alternative in which I can fork and then clone every single easystats package, but this is of course completely overkill and unmaintainable for my brain. I imagine this is what you have setup @DominiqueMakowski? 🤯

One issue here is in debugging problems users that users report on the master branches, or for creating an isolated environment for developing a package (and not ruining my existing install of easystats — e.g., development of insight). I've been thinking about setting up an [renv](https://github.com/rstudio/renv), but given the cross dependencies, but I'm not sure if this is overkill again, and I'm sure there's a better solution. 😫

--------------------------------------------------

TL;DR: How do you set up your easystats development environment? How isolate production from development? I think whatever answers emerge here would be worth documenting a best practice to help contributors come up to speed.

_I landed at this because tonight I landed at a bug in report which effectively was caused by an older development version of parameters, and I just reinstalled everything from scratch._

Best all 😊

**Comment by humanfactors** [comment]
Hey @DominiqueMakowski it's been a while and I've restructured my thinking of this question.

In short, how do you isolate the development version of each package you are working on? Do you use `renv`? If so, do you have separate environments for each package, or some chaotic web of all of them? Or is there no isolation and your forks _are_ easystats... so to speak...

**Comment by DominiqueMakowski** [comment]
ooooops, it seems like I've missed this issue altogether, my bad!! 😬 

- Issue of installation: yes this ~has been~ is an issue, because of the fast development pace and because some of the packages are still not on CRAN (but getting there). Basically, as all packages depend on one another, it pretty much requires to have everything updated at all times, which can be annoying. I am confident that this issue will disappear further as the ecosystem gets stabilized. 

> I'm guessing there's an alternative in which I can fork and then clone every single easystats package

That's a good question. I don't know how the others work and would be curious to hear. AFAIK, what I do is the following: 

First, I have github desktop (never managed to get used to git 💁‍♂), and a folder called "easystats" (that is in my dropbox, so it gets automatically updated *at least* across all the different machines I am using). In this easystasts folder, I have cloned all the easystats repos, so I have a folder for each package.

Now, for instance, recently I wanted to update parameters to do some edits. So what I do is I open github desktop, "fetch" the corresponding repo, which updates my local folder from github. Then I open `parameters.Rproj` and click on "build" which basically installs/updates the package in my R. 

However, as it sometimes happens, this threw an error, saying "parameters" (the new version I was trying to install) requires insight > 0.8.1, which I didn't have (because Daniel updated the requirements and I didn't update insight). So I closed R, fetched/updated the insight repo, built the updated insight, and then built parameters again.

However, if I hadn't used/updated for a certain period of time, I usually do a global update, so I fetch all the folders from github (note that as I am usually on the dev branch, I make sure to click on "update from master" because some ~crazy~ people commit directly to master, which results in the dev being behind master), and then re-build all the packages, in the following order: insight, bayestestR, performance, parameters, and the rest.

It might sound like a tedious process, but either I got used to it, or it's not that tedious once you mastered the github fetching -> building sequence.

I don't know if there's any better or simpler way of getting around that, though, curious to see how others do @strengejacke @mattansb @IndrajeetPatil

**Comment by strengejacke** [comment]
I would recommend following, if you want to use the easyststs package:

1) install each single package from GitHub, starting with insight, bayestestR, parameters, performance, effectsize, modelbased, see, then correlation and report.

2) now it _should_ be possible to install _easystats_ from GitHub.

I'm using GitHub desktop and git integration in RStudio. Usually, I open GitHub desktop and leave it open in the background for a while. After some minutes, you see which packages have new commits, so you can pull all changes.

As a user, you probably will never need to clone the repositories. Once you've installed easyststs, you can simply use `install_easystats_latest()` to get the latest updates from GitHub. Or you do nothing and wait for official CRAN updates. When you load the library (`library(easyststs)`), you will be informed which packages are up to date and which have newer versions on CRAN.

**Comment by strengejacke** [comment]
Btw, once you have installed easyststs, and want to contribute, simply update all packages from GitHub as described above, then fork the package you need, and simply file an PR. I think this is the easiest way of contributing of you have no direct write access.

**Comment by strengejacke** [comment]
> How isolate production from development?

I don't do that. Stable versions on CRAN, development versions on GitHub. Though the GitHub versions of the packages I maintain are usually "stable" ;-) , it's just that they probably depend on changes in the other GitHub repos. But, as I said, simply use install_easystats_latest() after starting RStudio, and it should be ok.

**Comment by humanfactors** [comment]
Thanks all for the feedback. It's more or less answered my question.

The issue with `install_easystats_latest` is that it's not transparent how it interacts with existing local git repositories, in terms of your R environment.

I was anticipating a more complex approach, where core development versions (i.e., forks of each package on dev branch) could be isolated in a virtual environment. This would ensure it is easier to actively develop a package, but then when I need to use a 'stable' version (i.e., to reproduce bugs reported in issues, or for work purposes), I could simply use the CRAN version. The trouble was/is that renv doesn't seem to have an intuitive way to deal with the complexities of co-dependent packages like you see in easy-stats. 

@DominiqueMakowski, a challenge that is also magnified for contributions from my perspective is that I don't feel totally comfortable working directly with the upstream easystats repository itself (in fact, I imagine everyone other than perhaps core team in theory should be on their own forks). It's not a big issue, but I do need to drop to the command line to pull upstream, or make my forks off upstream at time of editing.

I think what I will be working towards next instead (to expedite my development time) is simply have all `easystats` packages in a single directory, and have a script that can automatically ensure the local `dev` branch is up to date with `upstream/dev` (assuming no edits have been made).

**I will say my thoughts above are probably a reflection of my desire for perfection in face of complexity; rather than a pragmatic approach to GSD.**

I will close this for now, however if you do have another method @mattansb  @IndrajeetPatil feel free to reopen and comment of course.

**Comment by DominiqueMakowski** [comment]
> I was anticipating a more complex approach, where core development versions (i.e., forks of each package on dev branch) could be isolated in a virtual environment. This would ensure it is easier to actively develop a package alongside current/stable versions of all deps.

I think this is partly due to the fact that we have no background in software dev or computer science, to me "virtual envs" are black magic, and 'til now I managed by basically updating/installing everytime as any user could do. 

That being said, I'm open and curious to see what other approaches are, if they are indeed simpler (but as we said the process is quite straightforward and to be honest, now you need less and less to re-update all the easystats packages as they are getting more and more mature).

> a challenge that is also magnified for contributions from my perspective is that I don't feel totally comfortable working directly with the upstream easystats repository itself (in fact, I imagine everyone other than perhaps core team in theory should be on their own forks)

I can understand, in that case wouldn't it work to have your own fork and then updating it from dev, and the making a PR to dev?

> I think what I will be working towards next instead (to expedite my development time) is simply have all easystats packages in a single directory, and have a script that can automatically ensure the local dev branch is up to date with upstream/dev (assuming no edits have been made).

Let us know how it goes! But again, I wouldn't worry too much about breaking stuff and all, **that's what the dev branch is for**. And worst-case scenario we can revert. And worst-case worst-case scenario something is broken and no biggie, we are still in a development phase so it can happen ☺️

**Comment by strengejacke** [comment]
> The issue with install_easystats_latest is that it's not transparent how it interacts with existing local git repositories

Not at all. It just installs packages, like `install.packages()`, and it does not change any local repositories. It's just that when you work on _parameters_, which requires the latest GitHub version of _insight_ **installed** locally, you just run install_easystats_latest and can then work with the repositories if you want to make changes.

> and go back and start building my local versions, I assume it's overwriting the version called from install_easystats_latest?

Yes. But that's intended anyway.
